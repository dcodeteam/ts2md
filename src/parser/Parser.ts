import * as ts from "typescript";
import { ClassParseResult } from "./parsers/ClassParseResult";
import { NodeParseResult } from "./parsers/NodeParseResult";
import { InterfaceParseResult } from "./parsers/InterfaceParseResult";
import { VariableListParseResult } from "./parsers/VariableListParseResult";
import { FunctionParseResult } from "./parsers/FunctionParseResult";

export interface ParserResult {
  readonly nodes: Map<string, NodeParseResult>;
}

export class Parser {
  private readonly file: ts.SourceFile;

  public constructor(fileName: string, sourceText: string) {
    this.file = ts.createSourceFile(
      fileName,
      sourceText,
      ts.ScriptTarget.ESNext,
      true
    );
  }

  public parse(): ParserResult {
    const nodes = new Map<string, NodeParseResult>();

    this.file.statements.forEach(node => {
      if (ts.isClassDeclaration(node)) {
        const result = new ClassParseResult(node);

        nodes.set(result.id, result);
      }

      if (ts.isInterfaceDeclaration(node)) {
        const result = new InterfaceParseResult(node);

        nodes.set(result.id, result);
      }

      if (ts.isFunctionDeclaration(node)) {
        const result = new FunctionParseResult(node);

        nodes.set(result.id, result);
      }

      if (ts.isVariableStatement(node)) {
        const result = new VariableListParseResult(node);

        result.declarations.forEach(x => {
          nodes.set(x.id, x);
        });
      }
    });

    return { nodes };
  }
}
