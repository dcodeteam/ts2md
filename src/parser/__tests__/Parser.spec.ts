import * as fs from "fs";
import * as path from "path";
import { Parser } from "../Parser";

describe("Parser", () => {
  const fixturesDir = path.join(__dirname, "__fixtures__");

  fs.readdirSync(fixturesDir).forEach(file => {
    if (!file.endsWith(".ts")) {
      return;
    }

    test(file, () => {
      const filePath = path.join(fixturesDir, file);
      const sourceText = fs.readFileSync(filePath, "utf-8");

      const parser = new Parser(filePath, sourceText);

      expect(parser.parse()).toMatchSnapshot();
    });
  });
});
