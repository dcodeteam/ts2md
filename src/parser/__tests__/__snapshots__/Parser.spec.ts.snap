// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser class-docs.ts 1`] = `
Object {
  "nodes": Map {
    "Foo" => ClassParseResult {
      "defaultExported": false,
      "docs": Array [
        Object {
          "comment": "Foo class",
          "tag": "description",
        },
        Object {
          "comment": "foo",
          "tag": "see",
        },
        Object {
          "comment": "\`\`\`javascript
 * const foo = new Foo();
 * \`\`\`",
          "tag": "usage",
        },
      ],
      "exported": true,
      "extendedClass": null,
      "id": "Foo",
      "implementedInterfaces": Array [],
      "kind": 238,
    },
  },
}
`;

exports[`Parser class-extends.ts 1`] = `
Object {
  "nodes": Map {
    "Foo" => ClassParseResult {
      "defaultExported": false,
      "docs": Array [],
      "exported": false,
      "extendedClass": null,
      "id": "Foo",
      "implementedInterfaces": Array [],
      "kind": 238,
    },
    "Bar" => ClassParseResult {
      "defaultExported": false,
      "docs": Array [],
      "exported": true,
      "extendedClass": "Foo",
      "id": "Bar",
      "implementedInterfaces": Array [],
      "kind": 238,
    },
  },
}
`;

exports[`Parser class-implements.ts 1`] = `
Object {
  "nodes": Map {
    "Foo" => InterfaceParseResult {
      "defaultExported": false,
      "docs": Array [],
      "exported": false,
      "extendedInterfaces": Array [],
      "id": "Foo",
      "kind": 239,
    },
    "Bar" => InterfaceParseResult {
      "defaultExported": false,
      "docs": Array [],
      "exported": false,
      "extendedInterfaces": Array [],
      "id": "Bar",
      "kind": 239,
    },
    "Baz" => ClassParseResult {
      "defaultExported": false,
      "docs": Array [],
      "exported": true,
      "extendedClass": null,
      "id": "Baz",
      "implementedInterfaces": Array [
        "Foo",
        "Bar",
      ],
      "kind": 238,
    },
  },
}
`;

exports[`Parser class-mixed-extends-and-implements.ts 1`] = `
Object {
  "nodes": Map {
    "Foo" => InterfaceParseResult {
      "defaultExported": false,
      "docs": Array [],
      "exported": false,
      "extendedInterfaces": Array [],
      "id": "Foo",
      "kind": 239,
    },
    "Bar" => InterfaceParseResult {
      "defaultExported": false,
      "docs": Array [],
      "exported": false,
      "extendedInterfaces": Array [],
      "id": "Bar",
      "kind": 239,
    },
    "Baz" => ClassParseResult {
      "defaultExported": false,
      "docs": Array [],
      "exported": false,
      "extendedClass": null,
      "id": "Baz",
      "implementedInterfaces": Array [],
      "kind": 238,
    },
    "Quoz" => ClassParseResult {
      "defaultExported": false,
      "docs": Array [],
      "exported": true,
      "extendedClass": "Baz",
      "id": "Quoz",
      "implementedInterfaces": Array [
        "Foo",
        "Bar",
      ],
      "kind": 238,
    },
  },
}
`;

exports[`Parser export-class.ts 1`] = `
Object {
  "nodes": Map {
    "Foo" => ClassParseResult {
      "defaultExported": false,
      "docs": Array [],
      "exported": true,
      "extendedClass": null,
      "id": "Foo",
      "implementedInterfaces": Array [],
      "kind": 238,
    },
  },
}
`;

exports[`Parser export-default-class.ts 1`] = `
Object {
  "nodes": Map {
    "Foo" => ClassParseResult {
      "defaultExported": true,
      "docs": Array [],
      "exported": true,
      "extendedClass": null,
      "id": "Foo",
      "implementedInterfaces": Array [],
      "kind": 238,
    },
  },
}
`;

exports[`Parser export-default-function.ts 1`] = `
Object {
  "nodes": Map {
    "foo" => FunctionParseResult {
      "defaultExported": true,
      "docs": Array [],
      "exported": true,
      "id": "foo",
      "kind": 237,
    },
  },
}
`;

exports[`Parser export-default-interface.ts 1`] = `
Object {
  "nodes": Map {
    "Foo" => InterfaceParseResult {
      "defaultExported": true,
      "docs": Array [],
      "exported": true,
      "extendedInterfaces": Array [],
      "id": "Foo",
      "kind": 239,
    },
  },
}
`;

exports[`Parser export-function.ts 1`] = `
Object {
  "nodes": Map {
    "foo" => FunctionParseResult {
      "defaultExported": false,
      "docs": Array [],
      "exported": true,
      "id": "foo",
      "kind": 237,
    },
  },
}
`;

exports[`Parser export-interface.ts 1`] = `
Object {
  "nodes": Map {
    "Foo" => InterfaceParseResult {
      "defaultExported": false,
      "docs": Array [],
      "exported": true,
      "extendedInterfaces": Array [],
      "id": "Foo",
      "kind": 239,
    },
  },
}
`;

exports[`Parser export-variable.ts 1`] = `
Object {
  "nodes": Map {
    "FOO" => VariableParseResult {
      "defaultExported": false,
      "docs": Array [],
      "exported": true,
      "id": "FOO",
      "kind": 235,
      "type": "string",
    },
  },
}
`;

exports[`Parser interface-extends.ts 1`] = `
Object {
  "nodes": Map {
    "Foo" => InterfaceParseResult {
      "defaultExported": false,
      "docs": Array [],
      "exported": false,
      "extendedInterfaces": Array [],
      "id": "Foo",
      "kind": 239,
    },
    "Bar" => InterfaceParseResult {
      "defaultExported": false,
      "docs": Array [],
      "exported": false,
      "extendedInterfaces": Array [],
      "id": "Bar",
      "kind": 239,
    },
    "Baz" => InterfaceParseResult {
      "defaultExported": false,
      "docs": Array [],
      "exported": false,
      "extendedInterfaces": Array [
        "Foo",
        "Bar",
      ],
      "id": "Baz",
      "kind": 239,
    },
  },
}
`;

exports[`Parser variable-docs.ts 1`] = `
Object {
  "nodes": Map {
    "FOO" => VariableParseResult {
      "defaultExported": false,
      "docs": Array [
        Object {
          "comment": "Represents value of \`FOO\`.",
          "tag": "description",
        },
      ],
      "exported": true,
      "id": "FOO",
      "kind": 235,
      "type": "string",
    },
  },
}
`;

exports[`Parser variable-type.ts 1`] = `
Object {
  "nodes": Map {
    "FOO" => VariableParseResult {
      "defaultExported": false,
      "docs": Array [],
      "exported": true,
      "id": "FOO",
      "kind": 235,
      "type": "number",
    },
    "BAR" => VariableParseResult {
      "defaultExported": false,
      "docs": Array [],
      "exported": true,
      "id": "BAR",
      "kind": 235,
      "type": "number",
    },
    "BAZ" => VariableParseResult {
      "defaultExported": false,
      "docs": Array [],
      "exported": true,
      "id": "BAZ",
      "kind": 235,
      "type": "string",
    },
    "QUOZ" => VariableParseResult {
      "defaultExported": false,
      "docs": Array [],
      "exported": true,
      "id": "QUOZ",
      "kind": 235,
      "type": "string",
    },
  },
}
`;
